src/huggingface_hub/hf_api.py

```diff
@@ -1162,17 +1154,11 @@ def list_spaces(
             params.update({"datasets": datasets})
         if models is not None:
             params.update({"models": models})
-        r = requests.get(path, params=params, headers=headers)
-        hf_raise_for_status(r)
-        items = [SpaceInfo(**x) for x in r.json()]
-
-        # If pagination has been enabled server-side, older versions of `huggingface_hub`
-        # are deprecated as output is truncated.
-        _warn_if_truncated(
-            items, total_count=r.headers.get("X-Total-Count"), limit=limit
-        )
 
-        return items
+        data = paginate(path, params=params, headers=headers)
+        if limit is not None:
+            data = islice(data, limit)  # Do not iterate over all pages
```

Comment:
Shouldn't this logic be in `paginate` itself?

Here all the pages are first fetched from the API and only then is the slice done

Reply:
I hesitated but I think a "paginate" method should really stick to just paginate over requests, without stopping. Since it's a generator, only the requests that have to be made are actually made. The end of the generator is never unpiled (and therefore the vast majority of the calls are never triggered).

Reply:
Basically the `paginate` takes care of following the links, and the `islice` takes care of stopping when needed.

Makes it easy to work with the cases "limit is not set", "limit is set and LIMIT not hardcoded on server" "limit is set and greater than the LIMIT hardcoded by server (if any)" and "limit is set and lower than the LIMIT hardcoded by server (if any)".

Reply:
> the `islice` takes care of stopping when needed.

Oh ok! That's good then :)